File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/selection/GivenASavedMediaPlayerPlaybackEngineType/WhenGettingThePlaybackEngineType.java
package com.lasthopesoftware.bluewater.client.playback.engine.selection.GivenASavedMediaPlayerPlaybackEngineType;

import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;
import androidx.test.core.app.ApplicationProvider;

import com.lasthopesoftware.bluewater.ApplicationConstants;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.PlaybackEngineType;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.SelectedPlaybackEngineTypeAccess;
import com.lasthopesoftware.bluewater.shared.promises.extensions.FuturePromise;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(RobolectricTestRunner.class)
public class WhenGettingThePlaybackEngineType {

	private PlaybackEngineType playbackEngineType;

	@Before
	public void before() throws ExecutionException, InterruptedException {
		final SharedPreferences sharedPreferences = PreferenceManager
			.getDefaultSharedPreferences(ApplicationProvider.getApplicationContext());

		sharedPreferences.edit()
			.putString(ApplicationConstants.PreferenceConstants.playbackEngine, "MediaPlayer")
			.apply();

		final SelectedPlaybackEngineTypeAccess selectedPlaybackEngineTypeAccess =
			new SelectedPlaybackEngineTypeAccess(
				sharedPreferences,
				() -> new Promise<>(PlaybackEngineType.ExoPlayer));

		playbackEngineType = new FuturePromise<>(selectedPlaybackEngineTypeAccess
			.promiseSelectedPlaybackEngineType()).get();
	}

	@Test
	public void thenThePlaybackEngineTypeIsExoPlayer() {
		assertThat(playbackEngineType).isEqualTo(PlaybackEngineType.ExoPlayer);
	}

	@Test
	public void thenTheExoPlayerEngineIsTheSavedEngineType() {
		assertThat(PlaybackEngineType.valueOf(PreferenceManager
			.getDefaultSharedPreferences(RuntimeEnvironment.application)
			.getString(ApplicationConstants.PreferenceConstants.playbackEngine, "")))
			.isEqualTo(PlaybackEngineType.ExoPlayer);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/selection/GivenASavedMediaPlayerPlaybackEngineType/WhenGettingThePlaybackEngineType.kt
package com.lasthopesoftware.bluewater.client.playback.engine.selection.GivenASavedMediaPlayerPlaybackEngineType

import com.lasthopesoftware.AndroidContext
import com.lasthopesoftware.bluewater.client.playback.engine.selection.PlaybackEngineType
import com.lasthopesoftware.bluewater.client.playback.engine.selection.SelectedPlaybackEngineTypeAccess
import com.lasthopesoftware.bluewater.settings.repository.ApplicationSettings
import com.lasthopesoftware.bluewater.settings.repository.access.HoldApplicationSettings
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class WhenGettingThePlaybackEngineType : AndroidContext() {
	companion object {
		private val applicationSettings = FakeApplicationSettings()
		private var playbackEngineType: PlaybackEngineType? = null
	}

    override fun before() {
		val selectedPlaybackEngineTypeAccess = SelectedPlaybackEngineTypeAccess(applicationSettings,
			mockk {
				every { promiseDefaultEngineType() } returns Promise(PlaybackEngineType.ExoPlayer)
			})

		playbackEngineType = selectedPlaybackEngineTypeAccess.promiseSelectedPlaybackEngineType().toExpiringFuture().get()
    }

    @Test
    fun thenThePlaybackEngineTypeIsExoPlayer() {
        assertThat(playbackEngineType).isEqualTo(PlaybackEngineType.ExoPlayer)
    }

    @Test
    fun thenTheExoPlayerEngineIsTheSavedEngineType() {
        assertThat(applicationSettings.promiseApplicationSettings().toExpiringFuture().get()?.playbackEngineTypeName)
            .isEqualTo(PlaybackEngineType.ExoPlayer.name)
    }

	private class FakeApplicationSettings : HoldApplicationSettings {
		private var applicationSettings = ApplicationSettings(playbackEngineTypeName = "MediaPlayer")

		override fun promiseApplicationSettings(): Promise<ApplicationSettings> =
			Promise(applicationSettings)

		override fun promiseUpdatedSettings(applicationSettings: ApplicationSettings): Promise<ApplicationSettings> {
			this.applicationSettings = applicationSettings
			return Promise(applicationSettings)
		}
	}
}
